lib/mongo/protocol.ex

```diff
@@ -90,6 +93,7 @@ defmodule Mongo.Protocol do
       {:ok, ssl_sock} ->
         {:ok, %{s | socket: {:ssl, ssl_sock}}}
       {:error, reason} ->
+        :gen_tcp.close(sock)
```

Comment:
Is it guaranteed to be a tcp socket at this point? The returned error has the `:ssl` tag on it.

Reply:
```
iex(1)> {:ok, sock} = :gen_tcp.connect('www.google.com', 80, [])
{:ok, #Port<0.3165>}
iex(2)> :ssl.connect(sock, [])
{:error, :ssl_not_started}
iex(3)> Enum.count(:erlang.ports())
7
iex(4)> :gen_tcp.close(sock)
:ok
iex(5)> Enum.count(:erlang.ports())
6
```
I can't find anything in the ssl:connect method doc, so I am not 100% sure about all cases. But it seems like the passed sock was not closed from what I have tested so far.

Reply:
Looks like the ssl app was not started in the above case, I experimented some more, most of the time, the socket seems to get closed on error. But calling `gen_tcp.close` never raises any error, even if the port is already closed.